src/hash_map.c

```diff
@@ -0,0 +1,562 @@
+    }
+  }
+
+  // Time to check if we've exceeded our Load Factor and grow the map if so
+  ret = hash_map_check_and_grow_map(hash_map);
+  // Just log on this failure because the map can continue to operate with degraded performance
+  RCUTILS_LOG_ERROR_EXPRESSION(RCUTILS_RET_OK != ret, "Failed to grow hash_map. Reason: %d", ret);
+
+  return RCUTILS_RET_OK;
+}
+
+rcutils_ret_t
+rcutils_hash_map_unset(rcutils_hash_map_t * hash_map, const void * key)
+{
+  HASH_MAP_VALIDATE_HASH_MAP(hash_map);
+  HASH_MAP_VALIDATE_ARGUMENT_NOT_NULL(key);
+
+  size_t key_hash = 0, map_index = 0, bucket_index = 0;
+  bool already_exists = false;
+  rcutils_hash_map_entry_t * entry = NULL;
+
+  already_exists = hash_map_find(hash_map, key, &key_hash, &map_index, &bucket_index, &entry);
+
+  if (!already_exists) {
+    // The entry isn't in the map, so just exit
+    return RCUTILS_RET_OK;
+  }
+
+  // Remove the entry from its bucket and deallocate it
+  rcutils_array_list_t * bucket = &(hash_map->impl->map[map_index]);
+  if (RCUTILS_RET_OK == rcutils_array_list_remove(bucket, bucket_index)) {
+    hash_map->impl->size--;
+    hash_map_deallocate_entry(&hash_map->impl->allocator, entry);
+  }
+
+  return RCUTILS_RET_OK;
+}
+
+bool
+rcutils_hash_map_key_exists(const rcutils_hash_map_t * hash_map, const void * key)
+{
+  // Verify input
+  if (NULL == hash_map) {
+    return false;
+  } else if (NULL == hash_map->impl) {
+    return false;
+  } else if (NULL == key) {
+    return false;
+  }
+
+  size_t key_hash = 0, map_index = 0, bucket_index = 0;
+  bool already_exists = false;
+  rcutils_hash_map_entry_t * entry = NULL;
+
+  already_exists = hash_map_find(hash_map, key, &key_hash, &map_index, &bucket_index, &entry);
+
+  return already_exists;
+}
+
+rcutils_ret_t
+rcutils_hash_map_get(const rcutils_hash_map_t * hash_map, const void * key, void * data)
+{
+  HASH_MAP_VALIDATE_HASH_MAP(hash_map);
+  HASH_MAP_VALIDATE_ARGUMENT_NOT_NULL(key);
+  HASH_MAP_VALIDATE_ARGUMENT_NOT_NULL(data);
+
+  size_t key_hash = 0, map_index = 0, bucket_index = 0;
+  bool already_exists = false;
+  rcutils_hash_map_entry_t * entry = NULL;
+
+  already_exists = hash_map_find(hash_map, key, &key_hash, &map_index, &bucket_index, &entry);
+
+  if (already_exists) {
+    memcpy(data, entry->value, hash_map->impl->data_size);
+    return RCUTILS_RET_OK;
+  }
+
+  return RCUTILS_RET_NOT_FOUND;
+}
+
+rcutils_ret_t
+rcutils_hash_map_get_next_key_and_data(
+  const rcutils_hash_map_t * hash_map,
+  const void * previous_key,
+  void * key,
+  void * data)
+{
+  HASH_MAP_VALIDATE_HASH_MAP(hash_map);
+  HASH_MAP_VALIDATE_ARGUMENT_NOT_NULL(key);
+  HASH_MAP_VALIDATE_ARGUMENT_NOT_NULL(data);
+
+  size_t key_hash = 0, map_index = 0, bucket_index = 0;
+  bool already_exists = false;
+  rcutils_hash_map_entry_t * entry = NULL;
+  rcutils_ret_t ret = RCUTILS_RET_OK;
+
+  if (NULL != previous_key) {
+    already_exists = hash_map_find(hash_map, key, &key_hash, &map_index, &bucket_index, &entry);
```

Comment:
@nburek meta^2: iteration looks expensive, I wonder if there's a cheaper way to track where did it left off the last time `rcutils_hash_map_get_next_key_and_data` was called.

Reply:
In the case where you're not modifying the map as you iterate through then this shouldn't be that much more expensive. The find should average out to be constant time. There would be some slight benefit in creating an iterator style method for this as it might not need to rehash the key each time. I didn't use the iterator approach for two reasons: 
1) My primary use case was to use this function when cleaning up the map and I would be modifying it by removing entries as I went. A simple iterator would be invalid if the map was modified in between. 
2) So far, the other data structures didn't look like they were using an iterator style approach. This function was more in line with what I was seeing from the other data structures. I didn't want to risk add complexity to this review by trying to set a precedent for iterators when it wasn't needed yet.